clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -714,6 +707,8 @@ private void close(KafkaChannel channel, boolean processOutstanding) {
 
     private void doClose(KafkaChannel channel, boolean notifyDisconnect) {
         try {
+            immediatelyConnectedKeys.remove(channel.selectionKey());
```

Comment:
I think we should clear `immediatelyConnectedKeys` at the end of each `poll()` as well.

Reply:
Good point. It looks like keys would remain in this set until disconnected.

Reply:
Thankfully `immediatelyConnectedKeys` are quite rare. This particular issue was introduced by the optimisation in how we clear keys in the following fix (which was also fixing an issue introduced by KIP-72):

https://github.com/apache/kafka/commit/9ed8bf273c6ee5569ed68b71782ec7ab947f9c93

It would be good to add a unit test for that too.

Reply:
Oops, that was actually my fault.

Reply:
Sure, I will try to add a test case.